src/layers.jl

```diff
@@ -91,28 +91,90 @@ end
 skip_identity(inplanes, outplanes, downsample) = skip_identity(inplanes, outplanes)
 
 """
-    addrelu(x, y)
+    mlpblock(planes, expansion_factor = 4, dropout = 0., dense = Dense)
+
+Feedforward block used in many vision transformer-like models.
+
+# Arguments
+- `planes`: Number of dimensions in the input and output.
+- `hidden_planes`: Number of dimensions in the intermediate layer.
+- `dropout`: Dropout rate.
+- `dense`: Type of dense layer to use in the feedforward block.
+- `activation`: Activation function to use.
+"""
+function mlpblock(planes, hidden_planes, dropout = 0., dense = Dense; activation = gelu)
+  Chain(dense(planes, hidden_planes, activation), Dropout(dropout),
+        dense(hidden_planes, planes, activation), Dropout(dropout))
+end
+
+"""
+    Attention{T}
 
-Convenience function for `(x, y) -> @. relu(x + y)`.
-Useful as the `connection` argument for [`resnet`](#).
-See also [`reluadd`](#).
+Self attention layer used by transformer models. Can be instantiated with a layer that produces
+the key, value and query vectors from the input.
 """
-addrelu(x, y) = @. relu(x + y)
+struct Attention{T}
+  qkv::T
+end
+
+Attention(in, out) = Attention(Dense(in, out * 3; bias = false))
+
+@functor Attention
+
+function (attn::Attention)(x::AbstractArray{T}) where T
+  q, k, v = chunk(attn.qkv(x), 3; dim = 1)
+  scale = convert(T, sqrt(size(q, 1)))
+  score = softmax(batched_mul(batched_transpose(q), k) / scale)
+  attention = batched_mul(v, score)
```

Comment:
This only accepts `x::AbstractArray{T,3}`, so I would write that. 

Many layers also accept input without the batch dimension. If this one wants to, it would need a separate implementation, without `batched_transpose`.

Also, does someone have a great answer for why `/ scale`, why not absorb this into the learnable parameters? I know everyone does this but am unsure whether it's a cargo cult.

Reply:
For large values of `scale`, without the division, the softmax will have outputs that are very large, leading to small gradients which will impact the learning capacity of the model

Reply:
This is what I've heard people say, but why is this better than scaling the weights that lead to `q`? In many implementations, that's a separate weight matrix.

Reply:
Without a constraint, I'm not sure `Wq` will naturally learn to scale down. Manually scaling down `Wq` in the forward pass will be tantamount to what's here. Unless you are asking why `scale` itself is not learned? I think the original attention paper mentions that it could be but `sqrt(d)` works well enough.

Reply:
I guess I'm wondering why the scale isn't applied once at initialisation, instead of every pass. In this implementation, maybe that's tricky, since there is only one big weight matrix. But why don't implementations which aren't like that do it at init?

Reply:
IIRC this is a pretty common feature in transformer models, so let's not try to optimize it for this PR.

Reply:
Yes, not suggesting to deviate from the wisdom of the ancestors in this PR...

But this question aside, the other point about the batch dimension. The present signature accepts any array but only works for N=3, it should be restricted. And perhaps there ought to be a method for N=2 since it's simple:
```
function (attn::Attention)(x::AbstractMatrix{T}) where T
    q, k, v = chunk(attn.qkv(x), 3; dim = 1)
    scale = convert(T, sqrt(size(q, 1)))
    v * softmax(q' * k / scale)
end
```

Reply:
> But this question aside, the other point about the batch dimension. The present signature accepts any array but only works for N=3, it should be restricted. 

The latest version of the code has that 👍🏽